### PR TITLE
feat(access): the friends access is now an access collection

### LINKED ISSUE #3391: Friends access does not inherit properly (Trac #3391)
_Original ticket http://trac.elgg.org/ticket/3391 on 41310617-08-02 by trac user slyhne, assigned to unknown._

(New summary by @mrclay: When someone comments on/likes a post with ACCESS_FRIENDS, the ACCESS_FRIENDS value is inherited, but not the owner. As a result, the users who can see the comment/like is a _different set of users_ than who can see the original post.)

Elgg version: 1.7

I have just experienced some strange access things regarding annotations on a entity (in this case a blog).

User A creates a blog, setting access to friends only.

User B comments on that blog. Comment gets an access of friends only

User C wants to read comments, but is not friend with User B, so User C can only read the blog, not see comment given by User B. User C gives a comment, which then can only be read by User A and and User C, not User B.

I can see from the code, that annotations allways inherit the access_id of the entity they belong to. That must give some issues when using anything else than public!

A fix could be to allways set annotations to public access, and then make sure they are only shown to those who have access to the entity they belong to.

### PR DESCRIPTION
**Pull request by jdalsem**
Fixes #3391, #5038
Refs #8073, #3872

- [x] upgrade docs
- [x] manual tests

### CODE REVIEW

**Comment by jdalsem** [comment]
will finish it this time...

File: views/default/admin/upgrades.php [hunk 1]
```diff
@@ -10,7 +10,7 @@
 	'subtype' => 'elgg_upgrade',
 	'limit' => false,
 	'private_setting_name' => 'is_completed',
-	'private_setting_value' => false,
+	'private_setting_value' => 0,
```
Comment (jdalsem) [inline views/default/admin/upgrades.php]: need to revert this

**Comment by jdalsem** [comment]
Manually tested an upgrade from 2.3 to master with some entities and annotations with friends access, they seem to be migrated correctly. Tests are passing so maybe someone else can have a look at the PR

File: engine/classes/Elgg/Upgrades/MigrateFriendsACL.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	 */
+	public function countItems() {
+		// users without a friends acl
+		$db = elgg()->getDb();
+		
+		return elgg_get_entities([
+			'type' => 'user',
+			'count' => true,
+			'wheres' => [
+				"e.guid NOT IN (
```
Comment (jeabakker) [inline engine/classes/Elgg/Upgrades/MigrateFriendsACL.php]: use QueryBuilder for `wheres`

File: engine/classes/Elgg/Upgrades/MigrateFriendsACL.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	public function run(Result $result, $offset) {
+		
+		$db = elgg()->getDb();
+		
+		$users = elgg_get_entities([
+			'type' => 'user',
+			'limit' => 1,
+			'offset' => $offset,
+			'wheres' => [
+				"e.guid NOT IN (
```
Comment (jeabakker) [inline engine/classes/Elgg/Upgrades/MigrateFriendsACL.php]: use QueryBuilder

**Review by jeabakker — CHANGES_REQUESTED** [review]

**Comment by jdalsem** [comment]
Replaced by #11506

**Pull request closed without merging**